gcc/config/riscv/riscv.cc

```diff
@@ -3728,6 +3728,65 @@ riscv_pass_fpr_pair (machine_mode mode, unsigned regno1,
 				   GEN_INT (offset2))));
 }
 
+static void
+riscv_vector_psabi_warnning ()
+{
+  warning (OPT_Wpsabi, "ABI for the vector type is currently in experimental"
+       "stage and may changes in the upcoming version of GCC.");
+}
+
+static bool
+riscv_arg_has_vector (const_tree type)
+{
+  bool is_vector = false;
+
+  switch (TREE_CODE (type))
+    {
+    case RECORD_TYPE:
+      if (!COMPLETE_TYPE_P (type))
+	break;
+
+      for (tree f = TYPE_FIELDS (type); f; f = DECL_CHAIN (f))
+	if (TREE_CODE (f) == FIELD_DECL)
+	  {
+	    if (!TYPE_P (TREE_TYPE (f)))
+	      break;
+
+	    if (VECTOR_TYPE_P (type))
+	      is_vector = true;
+	    else
+	      is_vector = riscv_arg_has_vector (TREE_TYPE (f));
+	  }
+
+      break;
+
+    case VECTOR_TYPE:
```

Comment:
Does it included fixed length vector type like `typedef int v4si __attribute__ ((vector_size (16)));`?

Reply:
Yes. It will treat the variable with this attribute as a vector. I think it's what we want. Is it?

Reply:
That should exclude since it not passed in vector reg

Reply:
Hi Kito. I want to filter out this case when the type has attribute `vector_size`. But found `vector_size` is for variable, which means we can't get it by `TYPE_ATTRIBUTES`. And haven't found the way to convert the `type` to `decl`. Do you have any idea to distinguish them ?

Reply:
Type size?

Reply:
Thanks Kito. Tested this API and found for `vint32m_1`, it will return `POLY_INT_CST` and for the other will return `INTEGER_CST`. It can work now with new test cases.